src/store/accounts/nonce.ts

```diff
@@ -23,13 +25,25 @@ export const currentAccountLocalNonceState = atom(get => {
   return get(localNonceState([address, network.url]));
 });
 
+export const lastApiNonce = atom(get => {
+  const principal = get(currentAccountStxAddressState);
+  if (!principal) return;
+
+  const { accountsApi } = get(apiClientState);
+  return accountsApi.getAccountNonces({ principal });
+});
```

Comment:
@beguene continuing here -- part of what makes it possible for this atom to become out of date is it's just a standard async atom, meaning it when it mounts (first use), it will fetch a value at that time. At any point when it's called again, it will not know to "refresh". 

That is the main driver for any of the `atomWithQuery` uses, as it the query atom only exposes the functionality that react-query has, but in the context of an atom. This gives us the ability to be forced to have to put a `useLastApiNonce` hook (a wrapper around a standard `react-query` hook) somewhere in the component tree, and then feed the response to some atom elsewhere.

There are few options I see:
- refactor the use of nonce throughout the application to not rely on any async atom state, and use a `useQuery` hook to pass the nonce into a more pure function
- refactor this async atom to use `atomWithQuery` like the ones that exist currently

Either way, you'll need to additionally trigger a "refetch" when the nonce is first called, eg when a user clicks "preview", it should initiate a refresh to ensure we have up-to-date data, or when the user selects "confirm" on the transaction signing popup.

Reply:
Thanks a lot for the detailed explanation, this makes sense. I will fix it.